core/imagedownloader.cpp

```diff
@@ -90,11 +87,29 @@ Thumbnailer::Thumbnail Thumbnailer::fetchImage(const QString &filename, const QS
 {
 	QUrl url = QUrl::fromUserInput(filename);
 	if (url.isLocalFile()) {
+		// We try to determine the type first by peeking into the file.
 		QString filename = url.toLocalFile();
-		if (isVideoFile(filename))
+		metadata md;
+		mediatype_t type = get_metadata(qPrintable(filename), &md);
+
+		// For io error or video, return early with the appropriate dummy-icon.
+		if (type == MEDIATYPE_IO_ERROR)
+			return { failImage, MEDIATYPE_IO_ERROR };
+		else if (type == MEDIATYPE_VIDEO)
 			return { videoImage, MEDIATYPE_VIDEO };
+
+		// Try if Qt can parse this image. If it does, use this as a thumbnail.
 		QImage thumb(filename);
-		return { thumb, thumb.isNull() ? MEDIATYPE_IO_ERROR : MEDIATYPE_PICTURE };
+		if (!thumb.isNull())
```

Comment:
have you observed overhead when doing this on a video file instead of an image:
`QImage thumb("someVideoFile.avi")`

Reply:
I added a 1.1 GB avi and it is super-fast. Quite to my embarrassment, the last time I complained about Qt's slow image loading the actual culprit was our own hashing. [Which just shows that the troubles with QMultiMedia during thumbnail-extraction could actually be my fault.]